### PR TITLE
feat: add a test runner action

### PR DESCRIPTION
**Pull request by lengau**
This adds a runner action for running automated tests and uploading the result. It installs the snap, runs a user-provided test script, and then reports the result of the testing in the call-for-testing issue. The test command may write to the [job summary](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-job-summary), which will then be included in the comment.

~~Example: [Codespell snap](https://github.com/lengau/codespell-snap/issues/6)~~ (old version)

Example 1: [Codespell](https://github.com/lengau/codespell-snap/issues/12)
Example B: [Ruff](https://github.com/lengau/ruff-snap/issues/32)

In this case, the test script does some basic tests (getting the version and the help text) and then checks out various repositories that use codespell and runs the snap against each of them. 

Limitations: This runs directly on the GitHub runner, so it only tests the amd64 version of the snap. Once github provides runners for other architectures though, it will be possible to expand this to run on those architectures using a matrix.

### CODE REVIEW

File: run-tests/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# snapcrafters/ci/setup-tests
```
Comment (jnsgruk) [inline run-tests/README.md:1]:
```suggestion
# snapcrafters/ci/run-tests
```

Should this be `run-tests`?

File: run-tests/action.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+          echo "Installing snap from '${{ inputs.channel }}'"
+          sudo snap install "${snap_name}" --channel "${{ inputs.channel }}"
+        fi
+
+    - name: Run tests
+      id: tests
+      shell: bash
+      run: |
+        # shellcheck disable=SC2288,SC2086,SC1083
+        ${{ inputs.test-script }}
```
Comment (jnsgruk) [inline run-tests/action.yaml:63]:
```suggestion
        ${{ inputs.test-script }} | tee test-output.txt
        echo "output=$(cat test-output.txt)" >> "$GITHUB_OUTPUT"
```

File: run-tests/action.yaml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      uses: actions/github-script@v7
+      with:
+        script: |
+          github.rest.issues.createComment({
+            issue_number: ${{ inputs.issue-number }},
+            owner: context.repo.owner,
+            repo: context.repo.repo,
+            body: `Automated testing ${{ steps.tests.conclusion }}.
+
+            Full logs are available at https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
```
Comment (jnsgruk) [inline run-tests/action.yaml:76]:
```suggestion
            body: `Automated testing ${{ steps.tests.conclusion }}.

            Full logs are available at https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}.
            
            ```
            ${{ steps.tests.outputs.output }}
            ```         
```

**Review by jnsgruk — COMMENTED** [review]
Overall this looks great, thanks! 

Could you take a look at my suggestions - I think it would be good to include the test output inline in the comment, too? If you could have a go at running that over your codespell snap so we can see the results, that would be great!

File: run-tests/action.yaml (hunk 2 above)
Comment (lengau) [reply run-tests/action.yaml]: I couldn't get multiline outputs to work nicely, so I used Github's step summary feature instead.

**Comment by lengau** [comment]
@jnsgruk After not being satisfied with the options for generating outputs, I decided to go with offering the user a way to write a summary. A user who wants to simply place stdout from their test script in the summary can do the following:

```yaml
# ...
jobs:
  test:
    name: 🗒️ Test snap
    needs: call-for-testing
    runs-on: ubuntu-latest
    steps:
      - name: 🗒️ Run tests
        uses: snapcrafters/ci/run-tests@main
        with:
          issue-number: ${{ needs.call-for-testing.outputs.issue-number }}
          github-token: ${{ secrets.GITHUB_TOKEN }}
          test-script: |
            echo "```" >> "${GITHUB_STEP_SUMMARY}"
            ./test.sh >>  "${GITHUB_STEP_SUMMARY}"
            echo "```" >> "${GITHUB_STEP_SUMMARY}"
```

**Comment by jnsgruk** [comment]
@lengau - not a fan of this, only because it feels error prone for setting it up. I don't want the barrier to be too high for contributors at any level to configure this.

You mentioned that you couldn't get multiline outputs working properly - that's likely just due to environment variable handling in bash.

How about a similar but subtly different approach - first in the run tests step:

```yaml
# ...
shell: bash
run: |
    ${{ inputs.test-script }} | tee test-output.txt
```

And then in the script phase where the output is written to the comment:

```yaml
- name: Comment on call for testing issue with test result
  if: success() || failure()
  uses: actions/github-script@v7
  with:
    script: |
      const fs = require('node:fs');
      const summary = fs.readFileSync('test-output.txt', 'utf8');
      github.rest.issues.createComment({
        issue_number: ${{ inputs.issue-number }},
        owner: context.repo.owner,
        repo: context.repo.repo,
        body: `Automated testing ${{ steps.tests.conclusion }}.
        
        Full logs are available at: https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
        
        <pre>
        ${summary}
        <pre>
        `
      })
```

**Comment by lengau** [comment]
@jnsgruk I ran into some issues with putting the full logs into the comment like that, so how about this compromise? The first and last 100 lines of the log, with the full log still linked.

File: run-tests/action.yaml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        script: |
+          const fs = require('node:fs');
+          const summary = fs.readFileSync('test-output.txt', 'utf8');
+          github.rest.issues.createComment({
+            issue_number: ${{ inputs.issue-number }},
+            owner: context.repo.owner,
+            repo: context.repo.repo,
+            body: `Automated testing ${{ steps.tests.conclusion }}.
+
+            Full logs are available at https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
```
Comment (jnsgruk) [inline run-tests/action.yaml:90]:
```suggestion
            Full logs are available at https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}. The first 100, and last 100 lines of the log are displayed in the Summary below.
```

File: run-tests/action.yaml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        # shellcheck disable=SC2288,SC2086,SC1083
+        ${{ inputs.test-script }} |& tee raw-output.txt
+
+    - name: Process tests
+      shell: bash
+      if: success() || failure()
+      run: |
+        if [[ $(wc -l raw-output.txt | cut -f1 -d' ') -gt 250 ]]; then
+          head --lines=100 raw-output.txt > test-output.txt
+          echo -e "..." >> test-output.txt
```
Comment (jnsgruk) [inline run-tests/action.yaml:71]:
```suggestion
          echo -e "\n\n(Logs truncated. See full logs at: https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }})\n\n" >> test-output.txt
```

**Review by jnsgruk — APPROVED** [review]
Some very minor adjustments to make it clearer what's happening, but otherwise LGTM! Thanks for the iteration!

**Review by jnsgruk — APPROVED** [review]

**Pull request merged**